onsp_co2_hr_expense_report/models/hr_expense.py

```diff
@@ -0,0 +1,106 @@
+from odoo import api, fields, models, _
+from typing import Union
+
+class HrExpense(models.Model):
+    _name = "hr.expense"
+    _inherit = ["hr.expense", "carbon.line.mixin"]
+
+    def is_debit(self) -> bool:
+        self.ensure_one()
+        return False
+
+    def is_credit(self) -> bool:
+        self.ensure_one()
+        return True
+
+    # --------------------------------------------
+    #                   COMPUTE
+    # --------------------------------------------
+
+    @api.depends("carbon_debit", "carbon_credit")
+    def _compute_carbon_balance(self):
+        for line in self:
+            line.carbon_balance = line.carbon_debit - line.carbon_credit
+
+    @api.depends(
+        'account_id.use_carbon_value',
+        'account_id.carbon_in_value',
+        'account_id.carbon_in_compute_method',
+        'account_id.carbon_in_uom_id',
+        'account_id.carbon_in_monetary_currency_id',
+
+        'product_id.carbon_in_value',
+        'product_id.carbon_in_compute_method',
+        'product_id.carbon_in_uom_id',
+        'product_id.carbon_in_monetary_currency_id',
+
+        'product_id.carbon_out_value',
+        'product_id.carbon_out_compute_method',
+        'product_id.carbon_out_uom_id',
+        'product_id.carbon_out_monetary_currency_id',
+
+        'quantity',
+        'total_amount',
+        'date'
+    )
+    def _compute_carbon_debt(self, force_compute: Union[bool, str, list[str]] = None):
+        super()._compute_carbon_debt(force_compute)
+
+
+    @api.model
+    def _get_states_to_auto_recompute(self) -> list[str]:
+        return ['draft']
+
+    @api.model
+    def _get_state_field_name(self) -> str:
+        return 'state'
+
+    @api.model
+    def _get_carbon_compute_possible_fields(self) -> list[str]:
+        return ['product_id', 'account_id']
+
+    def _get_carbon_compute_kwargs(self) -> dict:
+        res = super()._get_carbon_compute_kwargs()
+        res.update({
+            'carbon_type': 'in',
+            'date': self.date,
+            'from_currency_id': (self.company_id or self.env.company).currency_id,
+        })
+        return res
+
+    def _get_line_amount(self) -> float:
+        self.ensure_one()
+        return self.total_amount or 0.0
+
+    def can_use_account_id_carbon_value(self) -> bool:
+        self.ensure_one()
+        return self.account_id.use_carbon_value and self.account_id.has_valid_carbon_in_value()
+
+    def get_account_id_carbon_compute_values(self) -> dict:
+        self.ensure_one()
+        return {'carbon_type': 'in'}
+
+    # --- PRODUCT ---
+    def can_use_product_id_carbon_value(self) -> bool:
+        self.ensure_one()
+        return bool(self.product_id) and (
+            self.product_id.has_valid_carbon_in_value() or
+            self.product_id.has_valid_carbon_out_value()
+        )
```

Comment:
what if the product has only a carbon_out value? The carbon type for `hr.expense` is 'in' by default if I understand correctly your code, so we need a carbon_in value, right?